Unexpected behaviour of swarm

After adding label to a node which turn on maintenance mode on it, a few minutes after we have errors like this:
```
Error: Error response from docker API, status: 500, message: Error: image eclipseche/workspaceqbkump84detb4cdg_machinemvk8hw1pgqxs1hm1_user_dev-machine not found
```
during 5-7 minutes. Many workspaces failed to start.
More, this can happens on any node and doesn't depend on node under maintenance. So, seems, that adding maintenance label is a start point only, but real reason of fails is unknown. That image really exists on a node (different from node under maintenance) and that is we got 500 error code instead of 404.
Also take into account, that maintenance constraint is always added independently of nodes labels.

**Reproduction Steps:**
Add `com.codenvy.node-state=maintenance` label to docker service on any node.
Unfortunately, we cannot reproduce it locally or on staging, but it happens on production.
It is a subtask how to reproduce it on staging.

**Expected behavior:**
After adding maintenance label to a node, other nodes should work correctly.

**Observed behavior:**
Error on workspace starting with 500 code and `image not found` message on many nodes. But when image really doesn't exist we have 404 code and message `no such image`.

**Codenvy version:** 5.0.0-M8-SNAPSHOT
**OS and version:** CentOS Linux 7 (Core) 3.10.0-327.10.1.el7.x86_64
**Docker version:** Version: 1.11.1 API version: 1.23
**Swarm:** Server Version: swarm/1.2.3

```
2016-12-12 15:18:22,264[kspaceManager-0]  [ERROR] [o.e.c.a.w.s.WorkspaceManager 660]    - org.eclipse.che.api.core.ServerException: Start of environment wksp-z3i1 failed. Error: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

java.util.concurrent.ExecutionException: org.eclipse.che.api.core.ServerException: Start of environment wksp-z3i1 failed. Error: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

        at java.util.concurrent.FutureTask.report(FutureTask.java:122) ~[na:1.8.0_45]
        at java.util.concurrent.FutureTask.get(FutureTask.java:192) ~[na:1.8.0_45]
        at org.eclipse.che.api.workspace.server.WorkspaceManager.lambda$performAsyncStart$4(WorkspaceManager.java:639) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceManager$$Lambda$159/303085739.run(Unknown Source) [che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalRunnable.run(CopyThreadLocalRunnable.java:28) [che-core-commons-lang-5.0.0-M8.jar:5.0.0-M8]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) [na:1.8.0_45]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_45]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_45]
Caused by: org.eclipse.che.api.core.ServerException: Start of environment wksp-z3i1 failed. Error: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.doStart(WorkspaceRuntimes.java:665) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.lambda$startAsync$1(WorkspaceRuntimes.java:245) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes$$Lambda$158/1742872442.call(Unknown Source) ~[na:na]
        at org.eclipse.che.commons.lang.concurrent.CopyThreadLocalCallable.call(CopyThreadLocalCallable.java:30) ~[che-core-commons-lang-5.0.0-M8.jar:5.0.0-M8]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_45]
        ... 3 common frames omitted
Caused by: org.eclipse.che.api.core.ServerException: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:856) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.start(CheEnvironmentEngine.java:251) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.workspace.server.WorkspaceRuntimes.doStart(WorkspaceRuntimes.java:635) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        ... 7 common frames omitted
Caused by: org.eclipse.che.api.core.ServerException: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startInstance(CheEnvironmentEngine.java:947) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startEnvironmentQueue(CheEnvironmentEngine.java:791) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        ... 9 common frames omitted
Caused by: org.eclipse.che.api.core.ServerException: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

        at org.eclipse.che.plugin.docker.machine.MachineProviderImpl.startService(MachineProviderImpl.java:327) ~[che-plugin-docker-machine-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.lambda$startEnvironmentQueue$17(CheEnvironmentEngine.java:764) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine$$Lambda$175/29274215.startMachine(Unknown Source) ~[na:na]
        at org.eclipse.che.api.environment.server.CheEnvironmentEngine.startInstance(CheEnvironmentEngine.java:894) ~[che-core-api-workspace-5.0.0-M8.jar:5.0.0-M8]
        ... 10 common frames omitted
Caused by: org.eclipse.che.plugin.docker.client.exception.DockerException: Error response from docker API, status: 500, message: Error: image eclipse-che/workspacethz78fqgpz8lriqf_machinezfiny67wsy8xjyog_qatestio_dev-machine not found

        at org.eclipse.che.plugin.docker.client.DockerConnector.getDockerException(DockerConnector.java:1324) ~[che-plugin-docker-client-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.plugin.docker.client.DockerConnector.createContainer(DockerConnector.java:1075) ~[che-plugin-docker-client-5.0.0-M8.jar:5.0.0-M8]
        at com.codenvy.swarm.client.SwarmDockerConnector.createContainer(SwarmDockerConnector.java:94) ~[codenvy-machine-hosted-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.plugin.docker.machine.MachineProviderImpl.createContainer(MachineProviderImpl.java:525) ~[che-plugin-docker-machine-5.0.0-M8.jar:5.0.0-M8]
        at org.eclipse.che.plugin.docker.machine.MachineProviderImpl.startService(MachineProviderImpl.java:273) ~[che-plugin-docker-machine-5.0.0-M8.jar:5.0.0-M8]
        ... 13 common frames omitted
```

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
@@ -37,6 +37,7 @@ import java.io.IOException;
 import java.net.URI;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.regex.Pattern;
 
 import static com.google.common.base.MoreObjects.firstNonNull;
 import static com.google.common.primitives.Ints.tryParse;
@@ -54,6 +55,8 @@ public class SwarmDockerConnector extends DockerConnector {
     //TODO should it be done in other way?
     private final String                  nodeDaemonScheme;
     private final int                     nodeDescriptionLength;
+
+    private static final Pattern IMAGE_NOT_FOUND_BY_SWARM_ERROR_MESSAGE = Pattern.compile("^Error: image .*  not found$");
 
     @Inject
     public SwarmDockerConnector(DockerConnectorConfiguration connectorConfiguration,
@@ -94,7 +97,8 @@ public class SwarmDockerConnector extends DockerConnector {
             return super.createContainer(params);
         } catch (DockerException e) {
             // TODO fix this workaround. Is needed for https://github.com/codenvy/codenvy/issues/1215
-            if (e.getStatus() == 500 && e.getOriginError().endsWith("not found")) { // if swarm failed to see image
+            if (e.getStatus() == 500 &&
+                IMAGE_NOT_FOUND_BY_SWARM_ERROR_MESSAGE.matcher(e.getOriginError()).matches()) { // if swarm failed to see image
                 try {
                     Thread.sleep(5000);                   // wait a bit
                     return super.createContainer(params); // and retry after pause

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
@@ -93,6 +93,17 @@ public class SwarmDockerConnector extends DockerConnector {
         try {
             return super.createContainer(params);
         } catch (DockerException e) {
+            // TODO fix this workaround. Is needed for https://github.com/codenvy/codenvy/issues/1215
+            if (e.getStatus() == 500 && e.getOriginError().endsWith("not found")) { // if swarm failed to see image
+                try {
+                    Thread.sleep(5000);                   // wait a bit
+                    return super.createContainer(params); // and retry after pause
+                } catch (DockerException de) {
+                    throw decorateMessage(de);
+                } catch (InterruptedException ie) {
+                    throw decorateMessage(e);
+                }
+            }
             throw decorateMessage(e);
         }
     }

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
@@ -56,7 +56,7 @@ public class SwarmDockerConnector extends DockerConnector {
     private final String                  nodeDaemonScheme;
     private final int                     nodeDescriptionLength;
 
-    private static final Pattern IMAGE_NOT_FOUND_BY_SWARM_ERROR_MESSAGE = Pattern.compile("^Error: image .*  not found$");
+    private static final Pattern IMAGE_NOT_FOUND_BY_SWARM_ERROR_MESSAGE = Pattern.compile("^Error: image .* not found$");
 
     @Inject
     public SwarmDockerConnector(DockerConnectorConfiguration connectorConfiguration,

--- a/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
+++ b/plugins/plugin-hosted/codenvy-machine-hosted/src/main/java/com/codenvy/swarm/client/SwarmDockerConnector.java
@@ -56,7 +56,7 @@ public class SwarmDockerConnector extends DockerConnector {
     private final String                  nodeDaemonScheme;
     private final int                     nodeDescriptionLength;
 
-    private static final Pattern IMAGE_NOT_FOUND_BY_SWARM_ERROR_MESSAGE = Pattern.compile("^Error: image .* not found$");
+    private static final Pattern IMAGE_NOT_FOUND_BY_SWARM_ERROR_MESSAGE = Pattern.compile("^Error: image .* not found.*", Pattern.DOTALL);
 
     @Inject
     public SwarmDockerConnector(DockerConnectorConfiguration connectorConfiguration,

